### PR TITLE
[processor/deltatocumulative] limit tracked streams

### PR DESCRIPTION
**Pull request by tombrk**
**Description:** Adds a configurable upper limit to the number of tracked streams. This allows to introduce a upper bound to the memory usage.

**Testing:** Test case was added

**Documentation:** README was updated

### CODE REVIEW

File: internal/exp/metrics/streams/streams.go
Comment (RichieSams) [inline internal/exp/metrics/streams/streams.go:44]: Why return an error instead of a boolean for exists / no exist? Then the user has to go through the effort of testing the error type

File: internal/exp/metrics/streams/streams.go
Comment (RichieSams) [inline internal/exp/metrics/streams/streams.go:58]: If we just use a typedef, we can avoid the extra indirection in the methods. IE:

`type HashMap[T any] map[identity.Stream]T`. Then the methods don't have to do a double dereference (one for the struct itself, and one for `items`)

File: internal/exp/metrics/streams/streams.go
Comment (RichieSams) [inline internal/exp/metrics/streams/streams.go:45]: While I could see some implementations *could* potentially error during a store operation, the primary players (a raw map, and sync.Map() ) don't. So we're adding a lot of extra error handling for a potential future where someone wants it.

File: internal/exp/metrics/streams/streams.go
Comment (RichieSams) [inline internal/exp/metrics/streams/streams.go:46]: Similarly, for context, all of these operations *should* be fairly quick. I'm not sure if we'd need a context for say, context cancellation checking. I could potentially see it for context propagation of something like span contexts. But again, this is a fairly tiny operation. So wanting spans at that level or events at that level seems a bit extreme.

**Comment by RichieSams** [comment]
I realize this is still draft, and you're iterating. But when everything is ready, can we split the `internal/exp/metrics` changes into a separate PR?

File: processor/deltatocumulativeprocessor/README.md
Comment (RichieSams) [inline processor/deltatocumulativeprocessor/README.md:31]: The default is `0`, correct? (which is interpreted as off). Should we notate that in the docs?

**Comment by RichieSams** [comment]
Just a small comment. Otherwise, LGTM

File: processor/deltatocumulativeprocessor/internal/streams/limit.go
Comment (djaglowski) [inline processor/deltatocumulativeprocessor/internal/streams/limit.go:25]: Rather than error, should we be evicting a stream, maybe whichever is stalest?

File: processor/deltatocumulativeprocessor/config.go
Comment (djaglowski) [inline processor/deltatocumulativeprocessor/config.go:33]: We try not to leave TODOs in code unless there is an associated issue mentioned as well.

File: processor/deltatocumulativeprocessor/internal/streams/limit.go
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/streams/limit.go:36]: at this point they are errors imo, the consumer (processor) should decide what to do with them.
when https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31363 is merged, those will turn into metrics.
Until then I think they are best treated as errors

Comment (RichieSams) [reply processor/deltatocumulativeprocessor/internal/streams/limit.go:36]: Sure. But I guess my point is that eviction itself isn't an error. It's just something that happens. IE, eviction should not increment the `metrics_processed{error=true}` metric (I don't recall the correct name :P ).

Perhaps the eviction yes/no value should be returned as an additional return, instead of being convolved with the error status. Thoughts?

Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/streams/limit.go:36]: I agree eviction should not increment that specific counter, but rather be `streams_evicted`, so you can easily alert on that. Imo eviction should never be part of normal operations, as it signals a fairly serious capacity issue and thus is very close to an error scenario.

I think it's a not uncommon thing in Go to use the Error for non-fatal occurrences, as that's the whole point of having a value-based, easy to manipulate error system.

**Review by RichieSams — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
I'm merging without final approval from @djaglowski as he's not available this week, and I believe there's nothing here that can't be worked on via follow-up PRs.

**Pull request merged**
